src/read.c

```diff
@@ -5384,8 +5395,19 @@ avifResult avifDecoderReset(avifDecoder * decoder)
                 decoder->image->gainMap = avifGainMapCreate();
                 AVIF_CHECKERR(decoder->image->gainMap, AVIF_RESULT_OUT_OF_MEMORY);
             }
-            AVIF_CHECKERR(avifParseToneMappedImageBox(&decoder->image->gainMap->metadata, tmapData.data, tmapData.size, data->diag),
-                          AVIF_RESULT_INVALID_TONE_MAPPED_IMAGE);
+            avifBool isUnsupportedVersion;
+            const avifBool tmapParsingRes =
+                avifParseToneMappedImageBox(&decoder->image->gainMap->metadata, tmapData.data, tmapData.size, &isUnsupportedVersion, data->diag);
+            if (isUnsupportedVersion) {
+                // Forget about the gain map.
+                toneMappedImageItem = NULL;
+                mainItems[AVIF_ITEM_GAIN_MAP] = NULL;
+                avifGainMapDestroy(decoder->image->gainMap);
+                decoder->image->gainMap = NULL;
+                decoder->gainMapPresent = AVIF_FALSE;
+            } else {
+                AVIF_CHECKERR(tmapParsingRes, AVIF_RESULT_INVALID_TONE_MAPPED_IMAGE);
+            }
```

Comment:
Nit: I suggest restructuring the if-else statement so that we first check the return value `tmapParsingRes` and then check the output parameter `isUnsupportedVersion`. This requires nested if statements, but it makes it clearer that `isUnsupportedVersion` is effectively an error/reason code that we check only when the function fails.

```
            if (!tmapParsingRes) {
                if (isUnsupportedVersion) {
                    // Forget about the gain map.
                    toneMappedImageItem = NULL;
                    mainItems[AVIF_ITEM_GAIN_MAP] = NULL;
                    avifGainMapDestroy(decoder->image->gainMap);
                    decoder->image->gainMap = NULL;
                    decoder->gainMapPresent = AVIF_FALSE;
                } else {
                    return AVIF_RESULT_INVALID_TONE_MAPPED_IMAGE;
                }
            }
```

Note:  the current version also has a nested if statement hidden inside the `AVIF_CHECKERR` macro.